packages/puppeteer-core/src/bidi/core/Session.ts

```diff
@@ -0,0 +1,137 @@
+import type * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+import {EventEmitter} from '../../common/EventEmitter.js';
+import {debugError} from '../../common/util.js';
+import {throwIfDisposed} from '../../util/decorators.js';
+
+import {Browser} from './Browser.js';
+import type {Connection} from './Connection.js';
+
+export class Session extends EventEmitter<{
+  /**
+   * Emitted when the session has ended.
+   */
+  ended: {reason: string};
+}> {
+  // TODO: Update generator to include fully module
+  static readonly subscribeModules: string[] = [
+    'browsingContext',
+    'network',
+    'log',
+    'script',
+  ];
+  static readonly subscribeCdpEvents: Bidi.Cdp.EventNames[] = [
+    // Coverage
+    'cdp.Debugger.scriptParsed',
+    'cdp.CSS.styleSheetAdded',
+    'cdp.Runtime.executionContextsCleared',
+    // Tracing
+    'cdp.Tracing.tracingComplete',
+    // TODO: subscribe to all CDP events in the future.
+    'cdp.Network.requestWillBeSent',
+    'cdp.Debugger.scriptParsed',
+    'cdp.Page.screencastFrame',
+  ];
+
+  static async from(
+    connection: Connection,
+    capabilities: Bidi.Session.CapabilitiesRequest
+  ): Promise<Session> {
+    // Wait until the connection is ready.
+    while (true) {
+      const {result: ready} = await connection.send('session.status', {});
```

Comment:
This looks like an infinite loop if the limit of connections is exceeded. Can we use rxjs here and throw an error?

Reply:
By looking at https://w3c.github.io/webdriver/#dfn-readiness-state it might never be ready resulting an infinite loop.

Reply:
I've implemented backoffs.